### PR TITLE
Cad implementation

### PR DESCRIPTION
**Pull request by nafur**
This PR implements the interface added in #4773.

### CODE REVIEW

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    Trace("cdcac") << "Infeasible intervals for " << p << " " << sc
+                   << " 0 over " << d_assignment << std::endl;
+    auto intervals = infeasible_regions(p, d_assignment, sc);
+    for (const auto& i : intervals)
+    {
+      Trace("cdcac") << "-> " << i << std::endl;
+      std::vector<poly::Polynomial> l, u, m, d;
+      // TODO(Gereon): Factorize polynomials here.
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp]: A TODO should be usually accompanied by an issue.

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+      // Add all discriminants
+      addPolynomial(res, discriminant(p));
+
+      for (const auto& q : requiredCoefficients(p))
+      {
+        // Add all required coefficients
+        Trace("cdcac") << "Coeff of " << p << " -> " << q << std::endl;
+        addPolynomial(res, q);
+      }
+      // TODO(Gereon): Only add if p(s \times a) = a for some a <= l
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp]: Ditto

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+      for (const auto& q : i.d_lowerPolys)
+      {
+        if (p == q) continue;
+        Trace("cdcac") << "Resultant of " << p << " and " << q << " -> "
+                       << resultant(p, q) << std::endl;
+        addPolynomial(res, resultant(p, q));
+      }
+      // TODO(Gereon): Only add if p(s \times a) = a for some a >= u
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp]: Ditto

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+      {
+        if (p == q) continue;
+        Trace("cdcac") << "Resultant of " << p << " and " << q << " -> "
+                       << resultant(p, q) << std::endl;
+        addPolynomial(res, resultant(p, q));
+      }
+    }
+  }
+
+  for (std::size_t i = 0; i < intervals.size() - 1; ++i)
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp:190]: Cache size() call.

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    auto tmp = isolate_real_roots(p, d_assignment);
+    roots.insert(roots.end(), tmp.begin(), tmp.end());
+  }
+  roots.emplace_back(poly::Value::plus_infty());
+  std::sort(roots.begin(), roots.end());
+
+  // Now find the interval bounds
+  poly::Value lower;
+  poly::Value upper;
+  for (std::size_t i = 0; i < roots.size(); ++i)
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp:249]: Ditto

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    Trace("cdcac") << "\torigins: " << intervals.back().d_origins << std::endl;
+
+    // Remove redundant intervals
+    cleanIntervals(intervals);
+  }
+
+  Trace("cdcac") << "Returning intervals for "
+                 << d_variableOrdering[cur_variable] << ":" << std::endl;
+  for (const auto& i : intervals)
+    Trace("cdcac") << "-> " << i.d_interval << std::endl;
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp:367]:
```suggestion
  for (const auto& i : intervals)
  {
    Trace("cdcac") << "-> " << i.d_interval << std::endl;
  }
```

Also guard loop with `Trace.isOn("cdac")`.

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 7]
```diff
@@ -20,12 +20,32 @@
...
 namespace CVC4 {
 namespace theory {
 namespace arith {
 namespace nl {
 namespace cad {
 
+namespace {
+/** Removed duplicates from a vector. */
+template <typename T>
+void remove_duplicates(std::vector<T>& v)
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp:42]:
```suggestion
void removeDuplicates(std::vector<T>& v)
```

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 8]
```diff
@@ -54,32 +89,283 @@ const std::vector<poly::Variable>& CDCAC::getVariableOrdering() const
...
 {
-  return {};
+  Trace("cdcac") << "Looking for unsat cover for "
+                 << d_variableOrdering[cur_variable] << std::endl;
+  std::vector<CACInterval> intervals = getUnsatIntervals(cur_variable);
+  Trace("cdcac") << "Unsat intervals for " << d_variableOrdering[cur_variable]
+                 << ":" << std::endl;
+  for (const auto& i : intervals)
+    Trace("cdcac") << "-> " << i.d_interval << " from " << i.d_origins
+                   << std::endl;
```
Comment (mpreiner) [inline src/theory/arith/nl/cad/cdcac.cpp:317]: Guard with `Trace.isOn("cdac")`.

File: src/theory/arith/nl/cad/cdcac.cpp (hunk 1 above)
Comment (nafur) [reply src/theory/arith/nl/cad/cdcac.cpp]: Todo was removed. Factorization is not coming anytime soon...

File: src/theory/arith/nl/cad/cdcac.cpp (hunk 2 above)
Comment (nafur) [reply src/theory/arith/nl/cad/cdcac.cpp]: Issue was added.

File: src/theory/arith/nl/cad/cdcac.cpp (hunk 3 above)
Comment (nafur) [reply src/theory/arith/nl/cad/cdcac.cpp]: Issue was added.

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 9]
```diff
@@ -54,36 +91,300 @@ const std::vector<poly::Variable>& CDCAC::getVariableOrdering() const
...
+      // Add all discriminants
+      addPolynomial(res, discriminant(p));
+
+      for (const auto& q : requiredCoefficients(p))
+      {
+        // Add all required coefficients
+        Trace("cdcac") << "Coeff of " << p << " -> " << q << std::endl;
+        addPolynomial(res, q);
+      }
+      // TODO(#4814): Only add if p(s \times a) = 0 for some a <= l
```
Comment (ajreynol) [inline src/theory/arith/nl/cad/cdcac.cpp:173]: Can you update this reference to the project issue # and write `cvc4-projects #N`?

File: src/theory/arith/nl/cad/cdcac.cpp [hunk 10]
```diff
@@ -54,36 +91,300 @@ const std::vector<poly::Variable>& CDCAC::getVariableOrdering() const
...
+      for (const auto& q : i.d_lowerPolys)
+      {
+        if (p == q) continue;
+        Trace("cdcac") << "Resultant of " << p << " and " << q << " -> "
+                       << resultant(p, q) << std::endl;
+        addPolynomial(res, resultant(p, q));
+      }
+      // TODO(#4814): Only add if p(s \times a) = 0 for some a >= u
```
Comment (ajreynol) [inline src/theory/arith/nl/cad/cdcac.cpp:181]: Ditto.

**Comment by mpreiner** [comment]
LGTM, thanks for the fixes!

**Pull request merged**